### ISSUE TITLE
Segfault when reuse same model with same model-instance-id.

### ISSUE THREAD

**Issue by yoonsanghyeuk** [dlstreamer/dlstreamer#396]
Hello.

I've constructed a pipeline as follows:

```text
rtspsrc *** decodebin *** gvadetect --- sink

*** Not yet linked
--- Linked
```

When connecting the correct RTSP source, I link rtspsrc and decodebin with a new pad, and similarly link decodebin and gvadetect. To share the same model with another RTSP source, I've set the "model-instance-id" with the same value.

It works fine when using the correct source. 

However, if an incorrect source (such as a wrong URL) is connected, there is an issue. The state fails to change to "Playing". Set state to NULL, and release pipeline resources. 

After a connection failure, a segmentation fault occurs when connecting the correct source again with same model, same "model-instance-id.".

This issue seems to be related to setting the same "model-instance-id". Interestingly, when setting different "model-instance-id" values for the same model, no segmentation fault occurs.

I used "person-vehicle-bike-detection-2004" model like tutorial.

**Comment by yoonsanghyeuk** [comment]
Here is my sample code for test.

Run on ubuntu 22.04, dl-streamer 2023, i7-10700K.

```c++
#include <iostream>

#include "gst/gst.h"

typedef struct _PipelineData {
  GstElement *pipeline;
  GstElement *rtspSrc;
  GstElement *decodeBin;
  GstElement *detector;
  GstElement *classifier;
  GstElement *sink;

  GstBus *bus;
} PipelineData;

void RTSPPadAdded(GstElement *src, GstPad *new_pad, gpointer user_data) {
  auto *pd = (PipelineData *) user_data;

  GstPad *sink_pad = gst_element_get_static_pad(pd->decodeBin, "sink");

  GstCaps *new_pad_caps = gst_pad_get_current_caps(new_pad);
  GstStructure *new_pad_struct = gst_caps_get_structure(new_pad_caps, 0);
  const gchar *new_pad_type = gst_structure_get_name(new_pad_struct);
  const gchar *new_pad_media = gst_structure_get_string(new_pad_struct, "media");

  if (!g_str_equal(new_pad_media, "video")) {
    return;
  }

  if (gst_pad_is_linked(sink_pad)) {
    return;
  }

  gst_pad_link(new_pad, sink_pad);

  if (new_pad_caps != nullptr) {
    gst_caps_unref(new_pad_caps);
  }

  gst_object_unref(sink_pad);
}

void DecodeBinPadAdded(GstElement *src, GstPad *new_pad, gpointer user_data) {
  auto *pd = (PipelineData *) user_data;

  GstPad *sink_pad = gst_element_get_static_pad(pd->detector, "sink");

  if (gst_pad_is_linked(sink_pad)) {
    return;
  }

  gst_pad_link(new_pad, sink_pad);

  gst_object_unref(sink_pad);
}

PipelineData CreatePipeline(const std::string &url,
                            const std::string &model_path,
                            const std::string &model_proc_path,
                            const std::string &cls_path,
                            const std::string &cls_proc_path,
                            bool same_id) {
  PipelineData pd;

  pd.rtspSrc = gst_element_factory_make("rtspsrc", "rtspsrc");
  pd.decodeBin = gst_element_factory_make("decodebin", "decodebin");

  pd.detector = gst_element_factory_make("gvadetect", "gvadetect");
  pd.classifier = gst_element_factory_make("gvaclassify", "gvaclassify");

  GstElement *gvaWatermark = gst_element_factory_make("gvawatermark", "gvawatermark");
  GstElement *videoConvert = gst_element_factory_make("videoconvert", "videoconvert");

  pd.sink = gst_element_factory_make("xvimagesink", "xvimagesink");

  pd.pipeline = gst_pipeline_new("gst-pipeline");

  if (!pd.rtspSrc || !pd.decodeBin || !pd.detector || !pd.classifier || !gvaWatermark || !videoConvert || !pd.sink
      || !pd.pipeline) {
    throw std::logic_error("Create element error.");
  }

  gst_bin_add_many(GST_BIN(pd.pipeline),
                   pd.rtspSrc,
                   pd.decodeBin,
                   pd.detector,
                   pd.classifier,
                   gvaWatermark,
                   videoConvert,
                   pd.sink,
                   nullptr);

  if (!gst_element_link_many(pd.detector, pd.classifier, gvaWatermark, videoConvert, pd.sink, nullptr)) {
    gst_object_unref(pd.pipeline);
    throw std::logic_error("Link error.");
  }

  g_object_set(pd.rtspSrc, "location", url.c_str(), nullptr);

  // set id
  static int count = 0;
  gchar *d_model_id;
  gchar *c_model_id;

  if (same_id) {
    d_model_id = g_strdup_printf("detector-same-id");
    c_model_id = g_strdup_printf("classifier-same-id");
  } else {
    d_model_id = g_strdup_printf("detector-%d", count);
    c_model_id = g_strdup_printf("classifier-%d", count++);
  }

  g_object_set(pd.detector,
               "model",
               model_path.c_str(),
               "model-proc",
               model_proc_path.c_str(),
               "model-instance-id",
               d_model_id,
               nullptr);

  g_object_set(pd.classifier,
               "model",
               cls_path.c_str(),
               "model-proc",
               cls_proc_path.c_str(),
               "model-instance-id",
               c_model_id,
               nullptr);

  g_signal_connect(pd.rtspSrc, "pad-added", G_CALLBACK(RTSPPadAdded), &pd);
  g_signal_connect(pd.decodeBin, "pad-added", G_CALLBACK(DecodeBinPadAdded), &pd);

  g_free(d_model_id);
  g_free(c_model_id);

  return pd;
}

void RunPipeline(const std::string &url,
                 const std::string &model_path,
                 const std::string &model_proc_path,
                 const std::string &cls_path,
                 const std::string &cls_proc_path,
                 bool same_id) {
  PipelineData pd = CreatePipeline(url, model_path, model_proc_path, cls_path, cls_proc_path, same_id);

  gst_element_set_state(pd.pipeline, GST_STATE_PLAYING);

  GstState state, pending;
  GstStateChangeReturn ret = gst_element_get_state(pd.pipeline, &state, &pending, GST_CLOCK_TIME_NONE);

  // Return if failed to play.
  if (ret == GST_STATE_CHANGE_FAILURE) {
    g_printerr("State changing is failed.\n");
    gst_element_set_state(pd.pipeline, GST_STATE_NULL);
    gst_object_unref(pd.pipeline);
    return;
  }

  pd.bus = gst_element_get_bus(pd.pipeline);

  GstMessage *msg =
      gst_bus_timed_pop_filtered(pd.bus, GST_CLOCK_TIME_NONE, (GstMessageType) (GST_MESSAGE_ERROR | GST_MESSAGE_EOS));

  if (GST_MESSAGE_TYPE (msg) == GST_MESSAGE_ERROR) {
    g_error ("An error occurred! Re-run with the GST_DEBUG=*:WARN environment "
             "variable set for more details.");
  }

  gst_message_unref(msg);
  gst_object_unref(pd.bus);
  gst_element_set_state(pd.pipeline, GST_STATE_NULL);
  gst_object_unref(pd.pipeline);
}

int main(int argc, char *argv[]) {
  gst_init(&argc, &argv);

  std::string wrong_url = "rtsp://192.168.1.4:554/wrong/media.smp";
  std::string correct_url = "rtsp://192.168.1.5:554/high/media.smp";

  std::string model_path =
      "/model_path/zoo_models/intel/person-vehicle-bike-detection-2004/FP32/person-vehicle-bike-detection-2004.xml";
  std::string model_proc_path =
      "/opt/intel/dlstreamer/samples/gstreamer/model_proc/intel/person-vehicle-bike-detection-2004.json";

  std::string cls_path =
      "/model_path/zoo_models/intel/vehicle-attributes-recognition-barrier-0039/FP32/vehicle-attributes-recognition-barrier-0039.xml";
  std::string cls_proc =
      "/opt/intel/dlstreamer/samples/gstreamer/model_proc/intel/vehicle-attributes-recognition-barrier-0039.json";

  bool same_id = false;

  g_print("wrong url\n");
  RunPipeline(wrong_url, model_path, model_proc_path, cls_path, cls_proc, same_id);

  g_print("correct url\n");
  RunPipeline(correct_url, model_path, model_proc_path, cls_path, cls_proc, same_id);

  return 0;
}

```

**Comment by pmalatyn** [comment]
we just released 2024.0 version, it supports API 2.0 so I highly recommend to check it and in case if this problem is still valid please raise new issue
